components/service/glean/src/main/java/mozilla/components/service/glean/storages/EventsStorageEngine.kt

```diff
@@ -105,20 +128,21 @@ internal object EventsStorageEngine : StorageEngine {
             eventKeys
         }
 
-        val msSinceStart = monotonicElapsedMs - startTime
-        val event = RecordedEventData(metricData.category, metricData.name, msSinceStart, truncatedExtraKeys)
+        val jsonContent = serializeEvent(
+            metricData.category, metricData.name, monotonicElapsedMs, truncatedExtraKeys
+        ).toString()
 
         // Record a copy of the event in all the needed stores.
         synchronized(this) {
+            var needToSend = false
             for (storeName in metricData.getStorageNames()) {
-                val storeData = eventStores.getOrPut(storeName) { mutableListOf() }
-                storeData.add(event.copy())
-                if (storeData.size == Glean.configuration.maxEvents &&
-                    Glean.assembleAndSerializePing(storeName)) {
-                    // Queue background worker to upload the newly created ping
-                    PingUploadWorker.enqueueWorker()
+                if (writeEventToDisk(storeName, jsonContent)) {
```

Comment:
What if we stored things both in the `eventStores` and on disk: we could retain the old behaviour and delegate disk writes to a synchronized queue dispatched on a different thread (so that we don't block the `API` dispatcher for too long). 

Moreover, taking the log for as long as I/O would last sounds a bit too long :(